File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_CHATEVENT.cs
﻿using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_CHATEVENT : Message
    {
        public SID_CHATEVENT()
        {
            Id = (byte)MessageIds.SID_CHATEVENT;
            Buffer = new byte[26];
        }

        public SID_CHATEVENT(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_CHATEVENT;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            if (context.Direction == MessageDirection.ClientToServer)
                throw new GameProtocolViolationException(context.Client, $"Client is not allowed to send {MessageName(Id)}");

            var chatEvent = (ChatEvent)context.Arguments["chatEvent"];
            Buffer = chatEvent.ToByteArray(context.Client.ProtocolType.Type);

            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)}: {ChatEvent.EventIdToString(chatEvent.EventId)} ({4 + Buffer.Length:D} bytes)");

            return true;
        }

        public new byte[] ToByteArray(ProtocolType protocolType)
        {
            if (protocolType.IsChat())
            {
                return Buffer;
            }
            else
            {
                return base.ToByteArray(protocolType);
            }
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_CHATEVENT.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_CHATEVENT
        Inherits Message

        Public Sub New()
            Id = CByte(MessageIds.SID_CHATEVENT)
            Buffer = New Byte(25) {}
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_CHATEVENT)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            If context.Direction = MessageDirection.ClientToServer Then Throw New GameProtocolViolationException(context.Client, $"Client is not allowed to send {MessageName(Id)}")
            Dim chatEvent = CType(context.Arguments("chatEvent"), ChatEvent)
            Buffer = chatEvent.ToByteArray(context.Client.ProtocolType.Type)
            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)}: {ChatEvent.EventIdToString(chatEvent.EventId)} ({4 + Buffer.Length} bytes)")
            Return True
        End Function
    End Class
End Namespace
